### ISSUE TITLE
keras.io examples using model.stateless_call aren't setting training=True and aren't managing non_trainable state

### ISSUE THREAD

**Issue by matpalm** [keras-team/keras#20378]
both the examples 
* https://keras.io/guides/writing_a_custom_training_loop_in_jax/
* https://keras.io/guides/distributed_training_with_jax/ 

have a training loop that uses `stateless_call`, but neither sets `training=True` ( which means they are using the `training=False` behaviour )

i guess it's not strictly relevant to the first example, which has no interesting non_trainables, but the second example has batch norm and dropout.

additionally i don't see where this non_trainable state would be managed? 

* re: batchnorm; each replica might have the same params, but is getting a different batch, so would collect different batch norm stats after the stateless_call. these aren't being aggregated?
* re: dropout; the RNGKeys for dropout are in the non_trainable, but never change. shouldn't these be split after each forward pass?

**Comment by fchollet** [comment]
Good call, `training=True` should be passed in both cases.

> additionally i don't see where this non_trainable state would be managed?

`stateless_call` returns the updated value of the `non_trainable_variables` (as updated during the forward pass), including BN variables.

> re: dropout; the RNGKeys for dropout are in the non_trainable, but never change. shouldn't these be split after each forward pass?

Similarly, they get updated during the forward pass, if you pass the updated `non_trainable_variables` on the next call, the next forward pass sees different seeds. There is no concept of seed "splitting", seeds are inspectable integers and are always explicitly passed and updated, like any other non-trainable state variable. No special-casing of RNG seeds.

> re: batchnorm; each replica might have the same params, but is getting a different batch, so would collect different batch norm stats after the stateless_call. these aren't being aggregated?

The aggregation should happen automatically on the JAX side; JAX lets you write "local, single-process" style code and add sharding annotations which let the compiler know what to distribute and aggregate.

@qlzh727 wrote the example, maybe he can help answer your questions.

You can also try using `ModelParallel`, which should be somewhat easier: https://keras.io/guides/distribution/

**Comment by matpalm** [comment]
thanks @fchollet 

re: dropout & batchnorm i think i need to dive into `jax.sharding` to understand the aggregation ( last time i looked at this it was all `pmap` and `psum` ) i suppose it's all mean aggregation by default and that works for all batch norm stats? ( and seeds too? ) 

all good, i'll dive further. just got put off by the no training and thought i must be missing something fundamental

hope you're well
